polyply/src/ff_parser_sub.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2020 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+from vermouth.parser_utils import (
+    SectionLineParser, _tokenize, _substitute_macros, _parse_macro
+)
+from vermouth.ffinput import FFDirector, _get_atoms, _treat_atom_prefix
+
+class PolyplyFFParser(FFDirector):
+    '''
+    More strictly implements differences between .ff within polyply
+    and .ff within martinize2.
+    '''
+    @SectionLineParser.section_parser('moleculetype', 'edges',
+                                      negate=False, context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'non-edges',
+                                      negate=True, context_type='block')
+    @SectionLineParser.section_parser('link', 'edges',
+                                      negate=False, context_type='link')
+    @SectionLineParser.section_parser('link', 'non-edges',
+                                      negate=True, context_type='link')
+    @SectionLineParser.section_parser('modification', 'edges',
+                                      negate=False, context_type='modification')
+
+    def _edges(self, line, lineno=0, negate=False, context_type=''):
+        context = self.get_context(context_type)
+        tokens = collections.deque(_tokenize(line))
+        _parse_edges_new(tokens, context, context_type, negate=negate)
+
+def _parse_edges_new(tokens, context, context_type, negate):
+    """
+    Parse the edge directive and add edge-attributes when
+    required.
+    """
+    if negate and context_type != 'link':
+        raise IOError('The "non-edges" section is only valid in links.')
+
+    atoms = _get_atoms(tokens, natoms=2)
+    prefixed_atoms = []
+    for idx, atom in enumerate(atoms):
+        prefixed_reference, attributes = _treat_atom_prefix(*atom)
+        [attributes.pop(key, None) for key in ["atomname", "order", "resname"]]
```

Comment:
I don't like this use of list comprehensions, since the goal is not to make a list, but rather to remove several items from attributes. Instead use a full loop.